### PR TITLE
Faster GPU NMS operator

### PR DESCRIPTION
**Pull request by ptrendx**
## Description ##
This PR significantly improves the performance of `mx.sym.contrib.box_nms` on GPU.

@zhreshold @Jerryzcn FYI

Test cases:

| input size | topk | old time | new time | speedup |
| ------------ | ---- | --- | --- | --- |
| (1, 507, 5) | 507 | 0.78 ms | 0.14 ms | 5.57 |
| (1, 1875, 5) | 1875 | 2.34 ms | 0.22 ms | 10.63 |
| (1, 7500, 5) | 2400 | 4.07 ms | 0.41 ms | 9.93 |
| (1, 30000, 5) | 2400 | 4.17 ms | 0.41 ms | 10.17 |
| (1, 120000, 5) | 2400 | 5.81 ms | 0.47 ms | 12.36 |
| (1, 159882, 5) | 12001 | 10.48 ms | 1.17 ms | 8.96 |

### Essentials ###
Please feel free to remove inapplicable items for your PR.
- [x] Changes are complete (i.e. I finished coding on this PR)
- [ ] All changes have test coverage:
- Unit tests are added for small changes to verify correctness (e.g. adding a new operator)
- [ ] Code is well-documented:  
- For new C++ functions in header files, their functionalities and arguments are documented. 
- [x] To the my best knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change

## Comments ##

 - Currently the new function is called when there is no requirement for backward pass, but it is easily extendable to include that.

### CODE REVIEW

File: src/operator/contrib/bounding_box.cu [hunk 1]
```diff
@@ -24,14 +24,701 @@
...
+                                   0,
+                                   Stream<gpu>::GetStream(s)>>>(
+    data.dptr_, out->dptr_, workspace.scores,
+    data.shape_[1], data.shape_[2], N,
+    param.valid_thresh, param.id_index,
+    param.score_index, param.background_id);
+}
+
+template <bool check_topk, bool check_score, typename DType>
+__global__ void CompactData_kernel(const index_t* indices, const DType* source,
```
Comment (zhreshold) [inline src/operator/contrib/bounding_box.cu]: Mixed naming convention, not sure if cpplint complains about it.
Same for lot's of kernel functions in the rest of this file

File: src/operator/tensor/sort_op-inl.cuh [hunk 2]
```diff
@@ -246,15 +310,25 @@ inline typename std::enable_if<(std::is_same<KDType,mshadow::half::half_t>::valu
...
+  if (sorted_keys != nullptr) {
+    k = sorted_keys;
+    mshadow::Copy(*sorted_keys, keys, keys.stream_);
+  }
+  if (sorted_values != nullptr) {
+    v = sorted_values;
+    mshadow::Copy(*sorted_values, values, values.stream_);
+  }
+  const auto key_iter = thrust::device_pointer_cast(reinterpret_cast<half*>(k->dptr_));
+  const auto value_iter = thrust::device_pointer_cast(reinterpret_cast<half*>(v->dptr_));
```
[the remark this answers is not in the document]
Comment (ptrendx) [reply src/operator/tensor/sort_op-inl.cuh]: Yup, a leftover from rebase.

File: src/operator/contrib/bounding_box.cu (hunk 1 above)
Comment (ptrendx) [reply src/operator/contrib/bounding_box.cu]: Will change.

**Comment by ptrendx** [comment]
@zhreshold Do you have any other comments?

For the record - I think there are a few outstanding items to fully mark GPU implementation of NMS as "done", namely
 - handle the output required for backward in the new implementation
 - change the way sorting is done (currently sorting happens `batch size` times, and it turns out that for large topk it is not beneficial, the previous implementation that would always do 2 (larger) sorts was better in that regard.
That said, I will not be able to finish those 2 outstanding items before 1.6 code freeze and I believe performance improvements are enough to merge this PR and address those 2 points as part of the future PR.

File: src/operator/contrib/bounding_box.cu [hunk 3]
```diff
@@ -24,14 +24,701 @@
...
+  return (x + y - 1) / y;
+}
+
+inline index_t align(index_t x, index_t alignment) {
+  return ceil_div(x, alignment)  * alignment;
+}
+
+template <typename DType>
+__global__ void FilterAndPrepareAuxDataKernel(const DType* data, DType* out, DType* scores,
+                                               index_t num_elements_per_batch,
```
Comment (apeforest) [inline src/operator/contrib/bounding_box.cu]: nit: indentation alignment?

File: src/operator/contrib/bounding_box.cu [hunk 4]
```diff
@@ -24,14 +24,701 @@
...
+                                   Stream<gpu>::GetStream(s)>>>(
+    data.dptr_, out->dptr_, workspace.scores,
+    data.shape_[1], data.shape_[2], N,
+    param.valid_thresh, param.id_index,
+    param.score_index, param.background_id);
+}
+
+template <bool check_topk, bool check_score, typename DType>
+__global__ void CompactDataKernel(const index_t* indices, const DType* source,
+                                   DType* destination, const index_t topk,
```
Comment (apeforest) [inline src/operator/contrib/bounding_box.cu]: nit: indentation alignment?

**Pull request merged**